### ISSUE TITLE
Wordpress woocommerce checkout page breaking with standard ruleset on iOS Chrome or Safari. Need rule exclusions.

### ISSUE THREAD

**Issue by Danrancan** [coreruleset/coreruleset#3226]
### Your Environment

* CRS version: v3.3.4
* Paranoia level setting: PL1
* ModSecurity version: modsecurity-nginx-v1.0.3, libmodsecurity v3.0.8
* Web Server and version or cloud provider / CDN (e.g., Apache httpd 2.4.54): Raspberry Pi 4
* Operating System and version: Ubuntu Server 20.04

### Confirmation

[ x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by franbuehler** [comment]
Thank you for your report. 

Your first false positive is because this user-agent `jetmon/1.0 (Jetpack Site Uptime Monitor by WordPress.com)` doesn't send an request header `accept`. If this is your test, are you able to add this accept header?

Next false positive:
Multiple URL Encoding Detected
Here you can tune this with the following tuning rule after the CRS is included: 
SecRuleUpdateTargetById 920230 !ARGS:post_data

You can do the same with this argument and the rule 942430.

And then you can go ahead and go through your modsecurity_audit.log and proceed with the tuning.

I didn't use the debug logs. That's too much information. But you can work with the modsecurity_audit.log or maybe evenbetter the error.log.
I highly recommend the tuning tutorials written by our co-lead.
There you find a way on how to work with false positives, logs and how to tune them away.
I hope my first tuning rules are helpful and a start for you. Please have a look at the linked tutorials and ask again if you need more help or guidance.

**Comment by Danrancan** [comment]
> Your first false positive is because this user-agent `jetmon/1.0 (Jetpack Site Uptime Monitor by WordPress.com)` doesn't send a request header `accept`. If this is your test, are you able to add this accept header?

I'm not exactly sure how make my jetpack plugin send an accept header. I'm assuming this is what you are asking me to do correct? Do you have any tips on how I might go about getting it to send an accept header? Is this something I need to configure in nginx, the jetpack plugin files, css, wordpress, etc? I'm not even sure where to start or how to make jetpack send this accept header.

> Next false positive: `Multiple URL Encoding Detected` Here you can tune this with the following tuning rule after the CRS is included: `SecRuleUpdateTargetById 920230 !ARGS:post_data`

By "after the CRS is included", do you mean that I should add this line to `REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf` after my already included rule exclusions? I'm a bit confused about where to add this line.

> You can do the same with this argument and the rule `942430`.

So I should add `SecRuleUpdateTargetById 942430 !ARGS:post_data` to where exactly? Like what file should I be adding this to?

Thanks again for all of you help and input here. I am sorry if my questions sound too noob or lazy. I'm still new and learning.  Thanks.

**Comment by Danrancan** [comment]
@franbuehler , I updated my original post to include the current rule exclusions I have based upon my original logs. Should I remove these current rules in place of `SecRuleUpdateTargetById 942430 !ARGS:post_data` and `SecRuleUpdateTargetById 920230 !ARGS:post_data`?

**Comment by franbuehler** [comment]
Thanks for the update. So let's do it this way. Perfect!

If you can't control the headers of the monitoting user-agent, let's tune this rule too with something like:

```
# Don't enforce accept header for jetmon user-agent 
SecRule  REQUEST_HEADERS:User-Agent "@beginsWith jetmon" \
    "id:106x,\
    phase:1,\
    pass,\
    nolog,\
    ctl:ruleRemoveById=920300"
```

`REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf` is the right place for this type of tuning rules. That's good.

Your exclusion rules probably didn't work because the variable `REQUEST_URI` also contains the query string, aka the part after the ?. See here: https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-(v2.x)#user-content-REQUEST_URI. And because you do a `@streq`, a string equals, it doesn't match.
So you could change your rules from `REQUEST_URI` to `REQUEST_FILENAME` (also see reference manual).

Try if it works like this. 
And afterwards, if you want to tune more accurate (optional!), you can remove arguments from individual rules, instead of removing the whole rule with `ctl:ruleRemoveTargetById=942340:ARGS:post_data` instead of `ctl:ruleRemoveById=942340`. 
But I think a good first step is to get your application running without errors or greyed out parts...

I would like to add this tuning tutorial again because it's a really good help: https://www.netnea.com/cms/nginx-tutorial-8_handling-false-positives-modsecurity-core-rule-set/

**Comment by Danrancan** [comment]
Hey, sorry for the late reply. I've been pretty busy. Anyways, I got my exclusions to work properly with the following rules in my REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf file:
```
# Don't enforce accept header for jetpack's jetmon user-agent
# This fixes the grey checkout area on checkout when jetpack is enabled.
# No other rules are necessary to fix this.
SecRule  REQUEST_HEADERS:User-Agent "@beginsWith jetmon" \
    "id:1060,\
    phase:1,\
    pass,\
    nolog,\
    ctl:ruleRemoveById=920300"

# The above jetpack rules fixed the checkout issue (but not always). Keep this here for reference.
# referer: https://www.mcmo.is/checkout/
SecRule REQUEST_FILENAME "@streq /" \
    "id:1062,\
    phase:1,\
    pass,\
    nolog,\
    ctl:ruleRemoveById=920230,\
    ctl:ruleRemoveById=942430"
```
I havent had the time to fine tune these by removing arguments from individual rules, but I will get there when I have enough time to read the Handling False Positives Modsecurity Core rule set tutorials. Even though wordpress is now working, can we keep this open just Incase I need some extra help after reading the tutorial and trying to fine tune things?

Anyways, THANK YOU SO SO SO SO SO MUCH! You have greatly helped me in one of my last steps before my server becomes a golden image and ready for production. I've been tinkering with this for months, and didn't find a solution, so your help is EXTREMELY appreciated and I can't thank you enough!

Dan
